railgun/internal/proxy/clientgo_cached_proxy_resolver.go

```diff
@@ -23,15 +25,83 @@ import (
 
 // NewCacheBasedProxy will provide a new Proxy object. Note that this starts some background services
 // and the caller is thereafter responsible for closing the Proxy.StopCh.
-func NewCacheBasedProxy(ctx context.Context, logger logrus.FieldLogger, k8s client.Client, kongConfig sendconfig.Kong, ingressClassName string, processClasslessIngressV1Beta1 bool, processClasslessIngressV1 bool, processClasslessKongConsumer bool) Proxy {
-	return NewCacheBasedProxyWithStagger(ctx, logger, k8s, kongConfig, ingressClassName, processClasslessIngressV1Beta1, processClasslessIngressV1, processClasslessKongConsumer, DefaultStagger)
+func NewCacheBasedProxy(ctx context.Context,
+	logger logrus.FieldLogger,
+	k8s client.Client,
+	kongConfig sendconfig.Kong,
+	ingressClassName string,
+	processClasslessIngressV1Beta1, processClasslessIngressV1, processClasslessKongConsumer, enableReverseSync bool,
+) (Proxy, error) {
+	return NewCacheBasedProxyWithStagger(ctx, logger, k8s, kongConfig, ingressClassName, processClasslessIngressV1Beta1, processClasslessIngressV1, processClasslessKongConsumer, enableReverseSync, DefaultStagger)
 }
 
-func NewCacheBasedProxyWithStagger(ctx context.Context, logger logrus.FieldLogger, k8s client.Client, kongConfig sendconfig.Kong, ingressClassName string, processClasslessIngressV1Beta1 bool, processClasslessIngressV1 bool, processClasslessKongConsumer bool, stagger time.Duration) Proxy {
+func NewCacheBasedProxyWithStagger(ctx context.Context,
+	logger logrus.FieldLogger,
+	k8s client.Client,
+	kongConfig sendconfig.Kong,
+	ingressClassName string,
+	processClasslessIngressV1Beta1, processClasslessIngressV1, processClasslessKongConsumer, enableReverseSync bool,
+	stagger time.Duration,
+) (Proxy, error) {
+	// download the kong root configuration (and validate connectivity to the proxy API)
+	root, err := kongConfig.Client.Root(ctx)
```

Comment:
Sufficiency of reading `Root()` is mutually exclusive with the need of go-kong to read `<workspace>/kong` if workspaces are in use.

If the root endpoint always works with workspaces, the implementation of version getter in go-kong needs to be simplified.
If the opposite is true, the implementation here needs to be extended and take into account the case where GETting `/` will 403.

By the original implementation of KIC 1.x (below) my educated guess is that the former is true.
https://github.com/Kong/kubernetes-ingress-controller/blob/07a5d3088043fcbfa8fd45036e93c70693ef88d8/cli/ingress-controller/main.go#L255

Reply:
Good catch, 4947c8e5cae8e5c9e45909d668124f70fd626b5b updates this to work with a timeout as it did in 1.x :+1: